File: day3.rb
def shared_prefix_len(s1, s2)
  if s1.length == 0 || s2.length == 0 || s1[0] == s2[0]
    return 0
  end

  return 1 + shared_prefix_len(s1[1..], s2[1..])
end

def freqs(lines)
  raise RuntimeError, "empty array" if lines.empty?

  frequencies = Array.new(lines[0].length)

  lines.each do |line|
    line.each_with_index do |c, i|
      frequencies[i] ||= {"0" => 0, "1" => 0}
      frequencies[i][c] += 1
    end
  end

  frequencies
end

lines = $stdin.each_line.map do |line|
  line.strip.split("")
end.to_a

frequencies = freqs(lines)
gamma = frequencies.map do |x|
  if x["1"] >= x["0"]
    "1"
  else
    "0"
  end
end.join

epsilon = gamma.split("").map do |x|
  if x == "0"
    "1"
  else
    "0"
  end
end.join

# part 1
pp gamma.to_i(2) * epsilon.to_i(2)

def find_oxygen(lines, i = 0)
  raise "out of lines" if lines.empty?
  return lines.first if lines.length == 1

  fs = freqs(lines)
  lines = lines.select do |line|
    if fs[i]["1"] >= fs[i]["0"]
      line[i] == "1"
    else
      line[i] == "0"
    end
  end

  find_oxygen(lines, i + 1)
end

def find_co2(lines, i = 0)
  raise "out of lines" if lines.empty?
  return lines.first if lines.length == 1

  fs = freqs(lines)
  lines = lines.select do |line|
    if fs[i]["0"] <= fs[i]["1"]
      line[i] == "0"
    else
      line[i] == "1"
    end
  end

  find_co2(lines, i + 1)
end

oxygen = find_oxygen(lines).join
co2 = find_co2(lines).join

# part 2
pp oxygen.to_i(2) * co2.to_i(2)

File: day3.go
// I gave up on this one and rewrote day3 in ruby 😅
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func main() {
	s := bufio.NewScanner(os.Stdin)

	var lines []string

	for s.Scan() {
		if err := s.Err(); err != nil {
			log.Fatal(err)
		}

		lines = append(lines, s.Text())
	}

	ones := make([]int, len(lines[0]))
	zeros := make([]int, len(lines[0]))

	for _, line := range lines {
		for i, c := range line {
			switch c {
			case '0':
				zeros[i]++
			case '1':
				ones[i]++
			}
		}
	}

	var gamma uint64 = 0
	var mask uint64 = 0

	for i := range ones {
		mask = mask | (1 << i)
		bit := len(ones) - i - 1
		if ones[i] > zeros[i] {
			gamma = gamma | (1 << bit)
		}
	}

	var epsilon uint64 = (gamma ^ mask)

	fmt.Printf("gamma  =%012b\n", gamma)
	fmt.Printf("epsilon=%012b\n", epsilon)

	fmt.Println("part 1:", gamma*epsilon)

	oxygen := map[string]struct{}{}
	co2 := map[string]struct{}{}
	for _, l := range lines {
		oxygen[l] = struct{}{}
		co2[l] = struct{}{}
	}

	for i := range ones {
		expected := byte('0')
		if ones[i] >= zeros[i] {
			expected = '1'
		}

		for line := range oxygen {
			if line[i] != expected {
				delete(oxygen, line)
			}
		}

		if len(oxygen) <= 1 {
			break
		}
	}

	fmt.Printf("oxygen=%s", oxygen)
}

func firstKey(m map[string]struct{}) string {
	for k := range m {
		return k
	}
	panic("empty map")
}
